docs/architecture/labeling.md

```diff
@@ -0,0 +1,256 @@
+# Labeling
+
+## Requirements
+
+- User can label every top-level entity like Application, Runtime
+- User can search Application/Runtime by the specific label and its value
+- User can find out about all label keys used in the given tenant
+- User can define validation rules for a label with a given key but it is optional
+- There is one special label: **Scenarios**, that has additional requirements:
+    - Every object is labeled with **Scenarios**
+    - By default, **Scenarios** has one possible value: **default**
+    
+    
+## Proposed solution
+
+1. Application or Runtime can be tagged with a label that consists of **key** and **value**.
+1. A label can have related LabelDefinition when a user can define JSON schema for validation of values.
+Thanks to that our API is extremely simple and we can treat every label in the same way.
+Even UI developers can benefit from it because there are libraries that render proper JS component
+on the JSON schema basis. See [this library](https://github.com/json-editor/json-editor).
+
+2. For the given label you can provide only one value. Still, a value can be JSON array, depending on LabelDefintions schema.
+3. LabelDefinition is optional, but it will be created automatically if the user adds a label for which LabelDefinition does not exist.
+ 
+### API Changes
+
+1. Extend GraphQL API with  mutations and queries for a new type: **LabelDefinition** 
+
+```graphql
+type LabelDefinition {
+    key: String!
+    schema: String
```

Comment:
`CLOB`?

Reply:
Why do we use simple `String` value here and not `CLOB`? For all strings with schemas, specs etc. we used CLOB, so it would make sense to put it here as well.
We need to be consistent, so it's a good time to decide whether we go for CLOB, or we just abandon it completely (as String will fit as well).

Or, maybe we could use `Any` scalar - the output from GraphQL will be easier to parse as client won't have to parse JS object from string.